### PR TITLE
Improve Popup/Marker positioning logic

### LINKED ISSUE #4577: Popups automatically wrap with multiple world views
**mapbox-gl-js version**: 0.35.0 (does not exist on 0.34.0)

### Steps to Trigger Behavior

1. Create a popup under a mouse location on a zoom 0 map with multiple world views.
2. Mouse over a wrapped world map location.
3. Popup will automatically wrap.

### Possible culprit

It looks like this commit https://github.com/mapbox/mapbox-gl-js/pull/4555/files may be automatically be wrapping popup.setLngLat()

### Expected Behavior

The popup appears at unwrapped world location under the mouse cursor.

### Actual Behavior

The popup appears at wrapped world location

### PR DESCRIPTION
**Pull request by jfirebaugh**
* Wrap only on map move events; otherwise preserve requested LngLat
* Wrap in a way that best preserves object constancy; i.e. minimizes perceived changes in position

Fixes #4577; see discussion there.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page

### CODE REVIEW

**Comment by jfirebaugh** [comment]
Just realized all this map is totally wrong if the map is rotated: :man_facepalming:

File: src/ui/popup.js [hunk 1]
```diff
@@ -70,11 +71,11 @@ class Popup extends Evented {
      */
     addTo(map) {
         this._map = map;
-        this._map.on('move', this._update);
+        this._map.on('move', this._onMove);
         if (this.options.closeOnClick) {
             this._map.on('click', this._onClickClose);
         }
-        this._update();
+        this._update(false);
```
Comment (andrewharvey) [inline src/ui/popup.js]: The problem with this is a marker added within -180 to 180 long might not be shown on the map initially.

eg. in this view australia is world -1 and I've added a marker at long 170 but it doesn't show up until I move the map.
selection_561

Comment (jfirebaugh) [inline src/ui/popup.js]: Agreed -- I've changed this to always apply the wrapping. Only the logic based on prior position is bypassed on the update immediately following `setLatLng`.

File: src/ui/marker.js [hunk 2]
```diff
@@ -85,6 +86,7 @@ class Marker {
      */
     setLngLat(lnglat) {
         this._lngLat = LngLat.convert(lnglat);
+        this._wrappedLnLat = this._pos = null;
```
Comment (mourner) [inline src/ui/marker.js]: Let's use `_wrappedLngLat` for consistency.

File: src/ui/marker.js [hunk 3]
```diff
@@ -146,12 +148,26 @@ class Marker {
...
-        let pos = this._map.project(this._map.transform.renderWorldCopies ? this._lngLat.wrapToBestWorld(this._map.getCenter()) : this._lngLat)._add(this._offset);
-        // because rouding the coordinates at every `move` event causes stuttered zooming
+
+        if (!this._wrappedLnLat) {
+            this._wrappedLnLat = this._lngLat;
+        }
+
+        if (this._map.transform.renderWorldCopies) {
+            this._wrappedLnLat = smartWrap(this._wrappedLnLat, this._pos, this._map.transform);
+        }
```
Comment (mourner) [inline src/ui/marker.js]: Could be shorter:

```js
this._wrappedLngLat = this._map.transform.renderWorldCopies ? 
    smartWrap(this._lngLat, this._pos, this._map.transform) : this._lngLat;

File: src/ui/marker.js (hunk 2 above)
Comment (jfirebaugh) [inline src/ui/marker.js]: I decided to eliminate the variable entirely and just mutate `this._lngLat `.

**Comment by andrewharvey** [comment]
I tested the updated branch, looks good to me. :+1:

**Pull request merged**